src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -0,0 +1,132 @@
+        };
+    constructor(
+        private promiseFactory?: PromiseFactory,
+        private setTimeoutManager?: SetTimeoutManager,
+    ) {}
+
+    public registerEventToApplicationListener = (
+        eventType: string,
+        callback: ApplicationListener,
+    ): void => {
+        if (this.eventsToApplicationListenersMapping[eventType]) {
+            this.eventsToApplicationListenersMapping[eventType].push(callback);
+        } else {
+            this.eventsToApplicationListenersMapping[eventType] = [callback];
+        }
+        this.processDeferredEvents();
+    };
+
+    public registerAdapterListenerForEvent = (
+        event: Events.Event<any>,
+        eventType: string,
+    ): void => {
+        event.addListener(this.adapterListener(eventType));
+    };
+
+    private processDeferredEvents(): void {
+        this.deferredEvents = this.deferredEvents.filter(
+            event => this.processEvent(event.eventType, event.eventArgs, true) === false,
+        );
+    }
+
+    private async forwardEventToApplicationListener(
+        listener: ApplicationListener,
+        eventArgs: any[],
+    ): Promise<any> {
+        const result = listener(...eventArgs);
+        if (!!result && typeof result.then === 'function') {
+            if (this.promiseFactory) {
+                //wrapping application listener promise responses in a 4-minute promise race
+                //prevents the service worker going idle before a response is sent
+                await this.promiseFactory.timeout(result, 240000);
+            }
+        } else {
+            // it is possible that this is the result of a fire and forget listener
+            // wrap in 2-minute timeout to ensure it completes during service worker lifetime
+            if (this.setTimeoutManager) {
+                return this.setTimeoutManager.setTimeout(
+                    () => Promise.resolve(result),
+                    120000,
+                    `wrapper-timeout-${Date.now()}`,
+                );
+            }
+        }
+        return result;
+    }
+
+    private unregisterAdapterListener(event: Events.Event<any>, eventType: string) {
+        event.removeListener(this.adapterListener(eventType));
+    }
+
+    private unregisterApplicationListenerForEvent(
+        eventType: string,
+        listener: ApplicationListener,
+    ) {
+        const allListeners = this.eventsToApplicationListenersMapping[eventType];
+        const applicationListenerIndex = allListeners.findIndex(listener);
+        if (applicationListenerIndex < 0) {
+            console.error('no application listeners found for ', eventType);
+        } else {
+            if (allListeners.length === 1) {
+                delete this.eventsToApplicationListenersMapping[eventType];
+            } else {
+                this.eventsToApplicationListenersMapping[eventType] =
+                    allListeners.splice(applicationListenerIndex);
```

Comment:
I'm confused by this logic- if the listener we want to remove is in the middle of allListeners, won't this remove the following items in allListeners as well?

Reply:
oh my you're right! I need to add the argument for how many to remove (just the one!).

Reply:
I ended up including the remove function from lodash to make this cleaner